### PR TITLE
ipa-pwd-extop: allow enforcing 2FA-only over LDAP bind

### PR DESCRIPTION
**Pull request by abbra**
When authentication indicators were introduced in 2016, ipa-pwd-extop plugin gained ability to reject LDAP BIND when an LDAP client insists the authentication must use an OTP token. This is used by ipa-otpd to ensure Kerberos authentication using OTP method is done with at least two factors (the token and the password).

This enfrocement is only possible when an LDAP client sends the LDAP control. There are cases when LDAP clients cannot be configured to send a custom LDAP control during BIND operation. For these clients an LDAP BIND against an account that only has password and no valid token would succeed even if admins intend it to fail.

Ability to do LDAP BIND without a token was added to allow users to add their own OTP tokens securely. If administrators require full enforcement over LDAP BIND, it is cannot be achieved with LDAP without sending the LDAP control to do so.

Add IPA configuration string, EnforceLDAPOTP, to allow administrators to prevent LDAP BIND with a password only if user is required to have OTP tokens. With this configuration enabled, it will be not possible for users to add OTP token if one is missing, thus ensuring no user can authenticate without OTP and admins will have to add initial OTP tokens to users explicitly.

Fixes: https://pagure.io/freeipa/issue/5169

### CODE REVIEW

**Comment by abbra** [comment]
New test works fine. Temp commit results: http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/ced3bee0-d647-11ee-8b41-fa163ec210a9/report.html, `test_totp_ldap` test.

File: daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c [hunk 1]
```diff
@@ -252,6 +252,10 @@ static struct ipapwd_krbcfg *ipapwd_getConfig(void)
             tmparray = slapi_entry_attr_get_charray(config_entry,
                                                     "ipaConfigString");
             for (i = 0; tmparray && tmparray[i]; i++) {
+                if (strcasecmp(tmparray[i], "EnforceLDAPOTP") == 0) {
+                    config->enforce_ldap_otp = true;
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (abbra) [reply daemons/ipa-slapi-plugins/ipa-pwd-extop/common.c]: Thanks, I refactoring this bit to always run checks and skip NTLM part in FIPS case.

**Comment by flo-renaud** [comment]
@abbra thanks for the fix, works for me.

**Comment by flo-renaud** [comment]
master:

* 1d2897e3d7cc88c2c5698126ecb1e59fff396bbc ipa-pwd-extop: allow enforcing 2FA-only over LDAP bind
* 23b224d7ad2e90d03543a0001f9a83731a8a14a5 ipa-pwd-extop: add MFA note in case of a successful LDAP bind with OTP

**Comment by flo-renaud** [comment]
> @flo-renaud I think we will skip `ipa-4-9` backport for this feature. I removed the label.

Ok, I marked the ticket as fixed since all the backports are done.

**Comment by lachuzhnikov** [comment]
Hi, team
Adding EnforceLDAPOTP attribute by using `ipa config-mod --addattr ipaconfigstring=EnforceLDAPOTP` prevent technological account authentication used for external bind. 
Prevents password change for this accounts. 
Password authentication method enforced for technological accounts are ignored. 

How to permit password authentication for ldap bind accounts when EnforceLDAPOTP is enabled ?

**Pull request closed without merging**